flink-end-to-end-tests/flink-batch-sql-test/src/test/java/org/apache/flink/sql/tests/Generator.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.sql.tests;
+
+import org.apache.flink.connector.testframe.source.FromElementsSource;
+import org.apache.flink.table.data.GenericRowData;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.data.StringData;
+import org.apache.flink.table.data.TimestampData;
+import org.apache.flink.types.Row;
+
+import java.time.Instant;
+import java.time.LocalDateTime;
+import java.time.ZoneOffset;
+import java.util.ArrayList;
+import java.util.List;
+
+class Generator implements FromElementsSource.ElementsSupplier<RowData> {
+    private static final long serialVersionUID = -8455653458083514261L;
+    private final List<Row> elements;
+
+    static Generator create(
+            int numKeys, float rowsPerKeyAndSecond, int durationSeconds, int offsetSeconds) {
+        final int stepMs = (int) (1000 / rowsPerKeyAndSecond);
+        final long durationMs = durationSeconds * 1000L;
+        final long offsetMs = offsetSeconds * 2000L;
+        final List<Row> elements = new ArrayList<>();
+        int keyIndex = 0;
+        long ms = 0;
+        while (ms < durationMs) {
+            elements.add(createRow(keyIndex++, ms, offsetMs));
```

Comment:
The new implementation will consume more memory than the old one which will generate `row` iteratively on the fly. This could be a potential issue for large data volume batch tests.

Reply:
> I think it makes sense to add comment that only ALL_EXCHANGES_BLOCKING, ALL_EXCHANGES_HYBRID_FULL, and ALL_EXCHANGES_HYBRID_SELECTIVE are supported by the adaptive batch scheduler.

Definitely 👍 

> The new implementation will consume more memory than the old one which will generate row iteratively on the fly. This could be a potential issue for large data volume batch tests.

Yep, now all records are generated and then used during execution, while before records were generated on the fly.
It is still possible to have such an implementation, however I am going to add context:

The PR started with a port from bash to Java.
The first pass was easy, but included the use of many deprecated method.
With @XComp we opted for improving that in a second commit.
While solving the deprecation warning, I decided to re-use the `FromElementsSource` already implemented in the test utils. However, that source is meant to be fault-tolerant, so, it requires to be able to get any record produced by offset, hence the `List` of elements created is necessary.

Truth be told, for this test, no fault-tolerance mechanism is required. I could have another implementation of that without that strict requirement that can use records on the fly and forget about those.

@JingGe @XComp should I proceed? Thank you!

Reply:
Thanks for the reply. Batch processing does not rely on offset commonly. Would you please help me understand why the source should be fault-tolerant and requires getting record by offset for batch?

Reply:
@JingGe yeah, nothing related strictly to this case.

The `FromElementsSource` is actually generic and can be used and is used in the streaming case.

Here I am using it in batch table mode.

I am just reusing that as it is possible 👍

Reply:
In other words, I could have another implementation of the bounded source without any fault-tolerance guarantee 👍

Reply:
The original implementation was only for batch in `BatchSQLTestProgram`. This PR is for the migration that should not implicitly bring big change for the data generation that might cause performance issue later. In addition, the new implementation is still in the `flink-batch-sql-test` module which should be used only for batch. 
No sure if there are already similar generators in the stream-sql-test. If not, a new jira task could created and add the new implementation to the stream sql test.

Reply:
Jing actually has a good point on the memory consumption. I missed that one. 👍 We should continue generating the records on-the-fly to be closer to what the original test did.

Reply:
@JingGe @XComp 

Got it, the number of records is not huge, that's why I did not mention that 👍 
However, I understand your concerns as well 👍 

At this point I would write another generator as part of this PR.

However, I would provide that as part of test-utils rather than only confined to batch as other tests could benefit from that.

What do you guys think?

Reply:
Sounds great! Please feel free to create a follow up ticket and contribute the new generator with a new PR.

Reply:
> Got it, the number of records is not huge, that's why I did not mention that 👍

True - that's a valid point. I didn't check the number of elements as part of my last comment. I leave the decision up to you whether it's done in a new PR or part of this PR.

Reply:
@JingGe @XComp Thank you for the feedback!

@XComp I would merge this is as it is.

In the brackground I was already working on something similar, I will create another issue for adding a test source for batch tests and for Table API

Reply:
@JingGe any objections? The refactoring should be ok considering that the amount of data involved is quite low. The actual migration from bash to Java is also done in a separate commit which enables us to revert if we feel it's necessary. WDYT?

Reply:
I still think we should keep the on-the-fly implementation. But we can move forward for now and modify it once we have the performance issue.